Add the workflow article: publish 2021-08-09 (#5923)
GoogleChrome/web.dev: src/site/content/en/blog/vitals-tools-workflow/index.md

---
title: A performance-focused workflow based on Google tools
subhead: Combine Google tools to audit, improve and monitor your website effectively.
authors:
  - antoinebisch
  - gmimani

> date: 2021-08-05

Author: ```suggestion
date: 2021-08-09
```

Author: Updating date for publication on Monday

description: >
  With the growing importance of Core Web Vitals, site owners and developers increasingly focus on performance and key user experiences. Google provides many tools to help evaluate, optimize, and monitor pages, but users are often confused by the different sources of data and how to leverage them effectively. This article proposes a workflow combining several tools and clarifies where and how they make sense along the development process.
tags:
  - blog
  - web-vitals
  - performance
---

With the [Core Web Vitals](/vitals/) (CWV)
metrics now incorporated into
[Page experience signals](https://developers.google.com/search/docs/guides/page-experience#signals),
user experience optimization has increasingly become top of mind for website owners.
This article will suggest how the
[various performance-focused tools](/vitals-tools/)
that Google has made freely available can be combined to effectively evaluate
website health and identify pain points,
debug and optimize pages that need attention,
and continuously monitor and prevent regression.

<figure class="w-figure">
{% Img
src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/aQKhuF8mZNiQNrMXhpkm.png",

> alt="A diagram of the various steps outlined in this article. Evaluate website health and identify pain points, debug and optimize, monitor and continuous development.",

Reviewer: ```suggestion
alt="A diagram of the cycle outlined in this article: evaluate website health and identify pain points, debug and optimize, monitor and continuous development.",
```

Reviewer: In the final section before the conclusion, you mention the cyclical nature of this process in text. If I can see, I get this message three times: twice in a diagram and once in text. If I'm visually impaired I only get it one time, in the text.  This section corrects this.

width="800", height="465" %}
  <figcaption class="w-figcaption">
    Suggested workflow to optimize Core Web Vitals.
  </figcaption>
</figure>

CWV are evaluated from real user experiences on a website.
The more data that can be collected and analyzed based on real users,
the more successful website owners will be at optimizing their pages.
To this effect,
we strongly recommend implementing
[Real User Measurement](/vitals-field-measurement-best-practices/) (RUM)
monitoring in addition to this workflow.

> ## Step A: Evaluate website health and Identify pain points

Author: ```suggestion
## Step A: Evaluate website health and identify pain points
```

have impacted user experience over time
(overlaying your historical release data with the CrUX dashboard monthly results,
to understand what caused regressions).
- Narrow down on regression timelines, providing the focus for further investigation.

To know more about how to use the CruX dashboard, read
[Using the dashboard](/chrome-ux-report-data-studio-dashboard/#using-the-dashboard).

Once you get the overall picture,
the next step is to identify which pages or templates need the most attention.
These origin-level reports do not provide this insight,
but the Search Console does.

### Identify poor performing pages using Google Search Console

<figure class="w-figure">
{% Img
src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/w3CgACGvcnqjORHfWVNg.png",
alt="A screenshot of Core Web Vitals charts.",
width="800", height="639" %}
  <figcaption class="w-figcaption">
    Core Web Vitals report in the Search Console.
  </figcaption>
</figure>

[The Core Web Vitals report in Search Console](https://support.google.com/webmasters/answer/9205520)
provides a holistic site-wide view of your performance
and helps identify the types of pages that need the most attention.
Using this tool you can:

> - Identify the number of pages with a poor experience,
> need improvements, or a good experience,
> broken down by mobile and desktop views.

Reviewer: ```suggestion
- Identify the number of pages with a poor experience,
that need improvements, or a good experience,
broken down by mobile and desktop views.
```

Reviewer: Because this is already published, I've skipped over several things I could have requested changes to. 

I have another problem with this passage, specifically the phrase 'a good experience'. Am I identifying pages that need a better user experience or am I identifying pages that are already good?

- Get granular information on URL performance grouped by status,
metric, and URL groups (groups of similar web pages).

> Identify key templates of pages to improve based on the URL groups surfaced
> (each group contains URLs that provide a similar user experience).

Reviewer: ```suggestion
- Identify key templates of pages to improve based on the URL groups surfaced.
(Each group contains URLs that provide a similar user experience).
```

- Get detailed device-specific reports
showing how many URLs are in the different experience buckets for either mobile or desktop.

Here are a few things that you should consider when using Search Console:

- While other tools have free access,
the Search Console requires authorized user access.
If you do not have access to this tool,
reach out to the relevant team to get access or share insights.
- Only indexed URLs can appear in this report,
except for the coverage report where you can see URLs that were crawled but not indexed.
The URLs shown are the actual URLs for which data was recorded
(that is, data is not assigned only to a page's canonical URL,
as it is in most other reports).
- Data showcased in this report might look different from the data displayed in the CrUX Dashboard.
Search Console charts show status by URL,
whereas the CrUX Dashboard shows status by visit.
Therefore, if you have a few poorly performing URLs with a high volume of visits,
the Search Console chart will show only a few poor URLs,
while the CrUX Dashboard will show a high percentage of poor experiences
(visits broken down per metrics).

Analyzing your website using the Search Console tool

> narrows down your analysis to pages and page types that need the most attention.

Reviewer: ```suggestion
narrows your analysis to pages and page types that need the most attention.
```

This data covers the last 28 days of user experience.
With this tool, you can get the user experience metrics evaluated at the 75th percentile.
In other words, if 75% of user experiences are equal to,
or better than the threshold set for the metric,
then the performance is considered 'good.'
A URL passes the CWV assessment when 75% of all visits to the page
performed at or above CWV thresholds for FID, LCP, and CLS.
You can compare the URL-level data with origin-level data
to see if the individual URLs are better or worse than
the origin-level metrics and understand what is unique about the particular page.

#### Lab data, generated by Lighthouse

The overall Lighthouse score for the URL displays at the top of the page.
Below the field data section is the lab data section
that provides metrics collected from a real-time Lighthouse run
and surfaces opportunities for the page.

Running PSI for different types of pages—analyzing the
field data and opportunities—will provide you with
a high-level understanding of metrics and improvements to focus on.
With that information,
you are now ready to debug and optimize your pages which we will discuss in the next section.

> ## Step B: Debug and Optimize

Author: ```suggestion
## Step B: Debug and optimize
```

3. Use [Chrome DevTools](https://developer.chrome.com/docs/devtools/)
to debug performance issues and test code changes.

Once you have identified the pages and metrics that need attention,
the next questions to ask are:

- How can these pages be optimized?
- What are the low-hanging fruits to tackle immediately?
- What improvements require more planning?

### Uncover opportunities with Lighthouse

The first step to answer these questions is using Lighthouse.
It provides a snapshot of opportunities to improve the load performance
for specific pages based on a lab run.
You can access Lighthouse in several ways, including:

- Running
[Lighthouse from Chrome DevTools](https://developers.google.com/web/tools/lighthouse#devtools).
- At [web.dev/measure](/measure/).
- From [PageSpeed Insights](https://developers.google.com/speed/pagespeed/insights/).

Lighthouse identifies critical areas of optimization at load time
and by how much they could potentially improve the user experience.
Quickly identifying low-effort fixes helps to start the optimization process
before handling the bigger tasks.

While the metrics' absolute values may be interesting to look at,

> do keep in mind that they change based on the environment Lighthouse is running in

Reviewer: ```suggestion
keep in mind that they change based on the environment Lighthouse is running in
```

and many other factors,
such as the network conditions of the servers hosting the website and its third parties.
In addition,
Lighthouse simulates a user's experience on a mid-tier mobile on a slow-4G connection,
and might not represent your users' experience at the 75th percentile
(which the CWV metrics are based on).
The metrics are a good indication of what is going on in a website,
but we recommend focusing more on the optimization opportunities in the report.

{% Aside %}
Lighthouse analyzes pages at load time in a lab environment.

> While lab assessments of LCP are often useful to debug and optimize LCP in the field,

Reviewer: ```suggestion
While lab assessments of LCP are often useful to debugging and optimizing LCP in the field,
```

Author: would that not be "useful for debugging..."? "Useful to debugging" sounds odd.

One of the main benefits of using the extension over other tools
is that it measures the FID and CLS metrics during the interaction with the page,
in contrast with Lighthouse,
which only captures metrics until the page is loaded (and excludes FID).

{% Aside %}
The Web Vitals extension only works on Chrome desktop.
To assess the performance of a mobile site, use it in mobile emulation mode.
Also, switch to
[phone field data](/field-data-in-the-web-vitals-extension/#comparing-field-data-from-phones)
for more comparable results.
{% endAside %}

<figure class="w-figure w-screenshot">
{% Img
src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/Yc4KfWt47zYfcNRTcLq6.png",
alt="A screenshot from the Web Vitals extension.",
width="800", height="556" %}
  <figcaption class="w-figcaption">
    Real-time information surfaced by the Web Vitals extension.
  </figcaption>
</figure>

While these tools provide a lot of useful information,
a developer may want to dig deeper into the underlying issues
and test changes before releasing their code.
Chrome DevTools (CDT) is a great help for doing just that.

### Deep dive with Chrome DevTools

If you are new to CDT,

> you can get an overview of all the cool features—analyze runtime performance,
> optimize website speed, debug JavaScript,
> simulate mobile devices,
> override JavaScript and HTML, CSS,
> and JavaScript coverage—in

Reviewer: ```suggestion
you can get an overview of all the cool features—analyzing runtime performance,
optimizing website speed, debugging JavaScript,
simulating mobile devices,
overriding JavaScript and HTML, CSS,
and JavaScript coverage—in
```

Reviewer: Verbs used as nouns. They need to be gerunds.

[this post](https://developer.chrome.com/docs/devtools/).

Some features we recommend exploring in the context of CWV are the following:

- Load the page in the **Performance** tab to analyze the waterfall,
identify bottlenecks and
[observe the CWV flags](https://developer.chrome.com/blog/new-in-devtools-88/#web-vitals).

> - Activate the
> [show layout shift regions](https://developer.chrome.com/blog/new-in-devtools-77/#layoutshifts)
> in the

Author: ```suggestion
- Enable the
**[Layout Shift Regions](https://developer.chrome.com/blog/new-in-devtools-77/#layoutshifts)**
checkbox in the
```

> [Command Menu](https://developer.chrome.com/docs/devtools/command-menu/)
> and load the page in the **Network** tab to highlight layout shifts at load time
> (also useful while browsing the page).
> - Analyze code usage with the
> [Coverage](https://developer.chrome.com/docs/devtools/coverage/)
> tab and remove unused code to improve the LCP and FID metrics.
> - Enable the
> [Local Overrides](https://developer.chrome.com/blog/new-in-devtools-65/#overrides)
> to test code changes on a live page.
> Changes made through the Chrome DevTools will persist,

Author: ```suggestion
**[Command Menu](https://developer.chrome.com/docs/devtools/command-menu/)**
and load the page in the **Network** tab to highlight layout shifts at load time
(also useful while browsing the page).
- Analyze code usage with the
**[Coverage](https://developer.chrome.com/docs/devtools/coverage/)**
tab and remove unused code to improve the LCP and FID metrics.
- Enable the
**[Local Overrides](https://developer.chrome.com/blog/new-in-devtools-65/#overrides)**
to test code changes on a live page.
Changes made through Chrome DevTools will persist,
```

and you can perform audits such as those discussed above,
allowing for a quick analysis of the impact on user experience.

<figure class="w-figure w-screenshot">
{% Img
src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/81399qdVF2p7kVz5klUG.png",
alt="A screenshot from Chrome DevTools.",
width="484", height="306" %}
  <figcaption class="w-figcaption">
    The network tab in Chrome DevTools highlighting layout shift regions with the option activated.
  </figcaption>
</figure>

After going through the steps discussed so far,
the analyzed and fixed website should be in a better state.

To ensure the changes made had an actual positive impact on users,
you should carry out step A (Evaluate website health and Identify pain points)
of this article again.
Note that as CrUX data is aggregated over a 28-day period,
changes to user experience metrics may not be immediately apparent on PSI and CrUX.
This is why implementing RUM solutions is critical,
as they will surface metrics and validate fixes much faster.

User experience work is an iterative process
and to ensure new releases do not regress improvements made,
it is critical to have a robust monitoring and continuous development process in place.

> ## Step C: Monitor and Continuous Development

Author: ```suggestion
## Step C: Monitor and continuous development
```

      </tr>
      <tr>
        <td>Recommended for most accurate insights.
Default users can query 1TB of data per month.
Beyond that, the standard rates apply.</td>
        <td>Free within API quota.</td>
        <td>Free within API quota.</td>
        <td>Free.</td>
      </tr>
    </tbody>
  </table>
</div>

Collecting and reporting on metrics
is just one piece of adding governance to prevent regression
but it occurs after you have made changes to the site.
It would be ideal to identify potential pull requests that would degrade the performance ahead of code releases.
To this end, you can use
[Lighthouse-CI](https://web.dev/lighthouse-ci/)
with most continuous integration tools.
Using this,
you will be able to identify degrading PRs
and work on optimizations before code releases go live
and have an otherwise unexpected impact on users.

Keep in mind that the ideal setup to best identify areas of improvement
involves continuous capturing of your own RUM data by leveraging tools
and libraries such as
[web-vitals.js](https://github.com/GoogleChrome/web-vitals).

## Conclusion

<figure class="w-figure">
{% Img
src="image/ZDZVuXt6QqfXtxkpXcPGfnygYjd2/aQKhuF8mZNiQNrMXhpkm.png",

> alt="A diagram of the various steps outlined in this article. Evaluate website health and identify pain points, debug and optimize, monitor and continuous development.",

Reviewer: ```suggestion
alt="A diagram of the cycle outlined in this article: evaluate website health and identify pain points, debug and optimize, monitor and continuous development.",
```
